### DISCUSSION TITLE
Selection based on amount of audio(s)?

### DISCUSSION THREAD

**Discussion by WhoMI7** [CogentRedTester/mpv-sub-select#19] [Q&A]
I have it configured it in a way such that if a file contains alang: jap and eng both (Anime for example) then it's priority should be selecting japanese audio without sign and songs track (select full subtitle), and if I switch to eng audio then it will automatically switch to signs and songs track.

But the problem exists when I want to watch a english movie, say Black Panther for example. Now it only has english audio but now it shouldn't switch to forced (which only has translation to any language other than english) track.
I don't know if it's making sense or not. Maybe it's already possible?

What I am asking is: Is it possible that selection is made on the basis of number of audios?
If a file has only english audio then it would select english (Full Subtitle or SDH), not forced. 

Something like uosc's button config, where we can set button to appear on the basis of number of tracks present in a file(audio,has_audio,has_many_audio) but for subtitle selection. 
https://github.com/tomasklaen/uosc/blob/main/script-opts/uosc.conf#L54

**Comment by CogentRedTester** [comment]
If you don't want forced tracks to be selected you can use the condition field:

```json
{
    "alang": "eng",
    "slang": "eng",
    "condition": "not sub.forced"
}
```

If you actually want the number of tracks for some reason you can do it with a more complex condition:

```json
{
    "alang": "eng",
    "slang": "eng",
    "condition": "(function() local num_audio = 0; for _, track in ipairs(mp.get_property_native('track-list', {})) do num_audio = num_audio + (track.type == 'audio' and 1 or 0) end; return num_audio == 2; end)()"
}
```

You can see at the very end I have `num_audio == 2`. This means the preference will be selected if there are exactly two audio tracks. You can change the number to whatever you want, or change it to `num_audio < 2`, `num_audio > 2`, etc. The `track.type == 'audio'` section can be changed to `track.type == 'sub'` to count subtitles instead.

This is a pretty clunky way of writing a condition. I could be convinced to make it easier if anyone shows a further need to search the track list.

**Comment by WhoMI7** [comment]
I added this:
```
    {
        "alang": ["eng", "en"],
        "slang": ["eng", "en", "zxx", "und", "*"],
        "whitelist": [ "sign", "song", "default", "*" ],
	"condition": "not sub.forced"
    },
```
But it's still selecting forced track.
![image](https://user-images.githubusercontent.com/93109835/216748424-b3bb9246-245b-4f44-88d7-c620c276fd93.png)

Then I added this
```
{
    "alang": "eng",
    "slang": "eng",
    "condition": "(function() local num_audio = 0; for _, track in ipairs(mp.get_property_native('track-list', {})) do num_audio = num_audio + (track.type == 'audio' and 1 or 0) end; return num_audio == 2; end)()"
},
```
Outcome:
![image](https://user-images.githubusercontent.com/93109835/216748503-356d3985-6528-4637-8c91-be85a7473f7c.png)

Also there were 4 audios of same language:
![image](https://user-images.githubusercontent.com/93109835/216748556-184ad5e3-7777-4d6f-a11e-7a50b986ceec.png)
Maybe condition can be set on the basis of _different_ alang?
If alang jpn and eng are present then it will first select jpn and with non forced eng track (full subtitle) and if I switch to eng manually then it switches slang to signs and songs track (forced).
But if there is only n number eng of alang then it should just select non forced english track.

**Reply by CogentRedTester** [reply]
```json
    {
        "alang": ["eng", "en"],
        "slang": ["eng", "en", "zxx", "und", "*"],
        "whitelist": [ "sign", "song", "default", "*" ],
    },
    {
        "alang": ["eng", "en"],
        "slang": ["eng", "en", "zxx", "und", "*"],
        "condition": "not sub.forced"
    }, 
```

From the README:
> The script moves down the list of track preferences until any valid pair of audio and subtitle tracks are found. Once this happens the script immediately sets the subtitle track and terminates. If none of the tracks match then track selection is deferred to mpv.

**Comment by CogentRedTester** [comment]
>     {
>        "alang": ["eng", "en"],
>        "slang": ["eng", "en", "zxx", "und", "*"],
>        "whitelist": [ "sign", "song", "default", "*" ],
>        "condition": "not sub.forced"
>     },

This will not match to any of the subtitles in that file because none of them have `sign`, `song`, `default`, or `*` in their names.

>     {
>        "alang": "eng",
>        "slang": "eng",
>        "condition": "(function() local num_audio = 0; for _, track in ipairs(mp.get_property_native('track-list', {})) do num_audio = num_audio + (track.type == 'audio' and 1 or 0) end; return num_audio == 2; end)()"
>     },

This will not match to any subtitles in that file because that file has 4 subtitle tracks, but that condition, as you requested, returns true if there are exactly 2 subtitles tracks in the file. If there were two it would still match to the forced track because it comes first and you haven't included `not sub.forced` in the condition.

If you specifically want to match to the *second* sub track then do `"condition": "sub.id == 2"`. However I don't think you actually want to do this, you want to set a fallback preference that matches to english subtitles that aren't forced. So just put at the bottom of your file:

```json
    {
        "alang": "*",
        "slang": "eng?",
        "condition": "not sub.forced"
    }
```

**Comment by WhoMI7** [comment]
> > ```
> > {
> >    "alang": ["eng", "en"],
> >    "slang": ["eng", "en", "zxx", "und", "*"],
> >    "whitelist": [ "sign", "song", "default", "*" ],
> >    "condition": "not sub.forced"
> > },
> > ```
> 
> This will not match to any of the subtitles in that file because none of them have `sign`, `song`, `default`, or `*` in their names.

![image](https://user-images.githubusercontent.com/93109835/216803483-0b8d6435-9da8-447e-9c8a-c638b807f1e2.png)
But they are marked as default and forced.
Also, how do I give external sub priority here?
![image](https://user-images.githubusercontent.com/93109835/216803535-bff28a6b-278a-4b35-92fe-7cad990b0556.png)
by adding this in whitelist `"sub.external"`?

And how do I set priority of different flags?

![image](https://user-images.githubusercontent.com/93109835/216803701-7250a762-a3a5-4889-b9d7-504026faba3c.png)

**Reply by CogentRedTester** [reply]
> > > ```
> > > {
> > >    "alang": ["eng", "en"],
> > >    "slang": ["eng", "en", "zxx", "und", "*"],
> > >    "whitelist": [ "sign", "song", "default", "*" ],
> > >    "condition": "not sub.forced"
> > > },
> > > ```
> > 
> > 
> > This will not match to any of the subtitles in that file because none of them have `sign`, `song`, `default`, or `*` in their names.
> 
> ![image](https://user-images.githubusercontent.com/93109835/216803483-0b8d6435-9da8-447e-9c8a-c638b807f1e2.png)
> But they are marked as default and forced.

It has the default tag, but their *names* don't. Their names are `Forced`, `Track 2`, `SDH`, and `Track 4`. The whitelist and blacklist only match to the names.

> Also, how do I give external sub priority here?
>![image](https://user-images.githubusercontent.com/93109835/216803535-bff28a6b-278a-4b35-92fe-7cad990b0556.png)
> by adding this in whitelist `"sub.external"`?

No because the whitelist and blacklist only match names. You can use `sub.external` in the `condition` field.

> And how do I set priority of different flags?
> 
> ![image](https://user-images.githubusercontent.com/93109835/216803701-7250a762-a3a5-4889-b9d7-504026faba3c.png)

The flags available through `sub._` are shown in the [mpv manual](https://mpv.io/manual/master/#command-interface-track-list). You set priorities by making multiple entries in the json file. An earlier entry will have a higher priority.

**Reply by CogentRedTester** [reply]
@BhaturaGuy could you post the final json file you ended up with? It might help others who find this post in the future.

**Reply by WhoMI7** [reply]
This is the final result

```json
[
    {
        "alang": "j[ap]n?",
        "slang": [ "eng?", "und" ],
        "blacklist": [ "sign", "song", "forced", "default" ],
	"whitelist": [ "dialogue", "full", "english" ],
	"condition": "sub.external"
    },
    {
        "alang": "j[ap]n?",
        "slang": [ "eng?", "und" ],
        "blacklist": [ "sign", "song", "forced", "default" ],
	"whitelist": [ "dialogue", "full", "english" ]
    },
    {
        "alang": "eng?",
        "slang": ["eng?", "zxx", "und"],
        "whitelist": [ "sign", "song", "default" ],
	"condition": [ "sub.codec == 'ass'", "sub.external" ]
    },
    {
        "alang": "eng?",
        "slang": ["eng?", "zxx", "und"],
        "whitelist": [ "sign", "song", "default" ],
	"condition": "sub.codec == 'ass'"
    },
    {
        "alang": "eng?",
        "slang": [ "eng?", "und" ],
        "whitelist": [ "eng?", "track" ],
	"condition": [ "not sub.forced", "sub.default", "sub.codec == 'srt'", "sub.external" ]
    },
    {
        "alang": "eng?",
        "slang": [ "eng?", "und" ],
        "whitelist": [ "eng?", "track" ],
	"condition": [ "not sub.forced", "sub.default", "sub.codec == 'srt'" ]
    },
    {
        "alang": "*",
        "slang": [ "eng?", "und" ],
	"condition": [ "not sub.forced", "sub.default", "sub.codec == 'srt'", "sub.external" ]
    },
    {
        "alang": "*",
        "slang": [ "eng?", "und" ],
        "condition": [ "not sub.forced", "sub.default", "sub.codec == 'srt'" ]
    }	
]
```

Since movies and tv shows doesn't use `.ass` subs and anime(generally) doesn't use `.srt` subs it worked out pretty great.
But the external subs are preferred(if present) in the directory (default mpv behavior).

**Reply by CogentRedTester** [reply]
Just so you're aware conditions can't be placed into an array like that (though it's not a bad idea). If you want to evaluate multiple conditions you need to put them into a single string and combine them with `and` or `or`. So for example your final preference is currently `[ "not sub.forced", "sub.default", "sub.codec == 'srt'" ]`, but this won't work (or maybe it'll always return true) what you want to do instead is:

`"not sub.forced and sub.default and sub.codec == 'srt'"` which will require all three of those conditions to be true or alternatively `"not sub.forced or sub.default or sub.codec == 'srt'"` will match if any one of the three conditions is true.
